src/rgw/rgw_auth.h

```diff
@@ -0,0 +1,340 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+
+#ifndef CEPH_RGW_AUTH_H
+#define CEPH_RGW_AUTH_H
+
+#include <functional>
+#include <type_traits>
+
+#include "rgw_common.h"
+#include "rgw_keystone.h"
+
+#define RGW_USER_ANON_ID "anonymous"
+
+
+/* Load information about identity that will be used by RGWOp to authorize
+ * any operation that comes from an authenticated user. */
+class RGWIdentityApplier {
+public:
+  typedef std::map<std::string, int> aclspec_t;
+
+  virtual ~RGWIdentityApplier() {};
+
+  /* Translate the ACL provided in @aclspec into concrete permission set that
+   * can be used in authorization phase (particularly in verify_permission
+   * method of a given RGWOp).
+   *
+   * XXX: implementation is responsible for giving the real semantic to the
+   * items in @aclspec. That is, their meaning may depend on particular auth
+   * engine that was used. */
+  virtual uint32_t get_perms_from_aclspec(const aclspec_t& aclspec) const = 0;
+
+  /* Verify whether a given identity *can be treated as* an admin of
+   * the rgw_user (account in Swift's terminology) specified in @uid. */
+  virtual bool is_admin_of(const rgw_user& uid) const = 0;
+
+  /* Verify whether a given identity *is* the owner of the rgw_user
+  * (account in Swift's terminology) specified in @uid. */
+  virtual bool is_owner_of(const rgw_user& uid) const = 0;
+
+  /* Return the permission mask that is used to narrow down the set of
+   * operations allowed for a given identity. This method reflects the idea
+   * of subuser tied to RGWUserInfo. */
+  virtual uint32_t get_perm_mask() const = 0;
+
+  virtual bool is_anonymous() const final {
+    /* If the identity owns the anonymous account (rgw_user), it's considered
+     * the anonymous identity. */
+    return is_owner_of(rgw_user(RGW_USER_ANON_ID));
+  }
+
+  virtual std::string to_str() const = 0;
+};
+
+inline std::ostream& operator<<(std::ostream& out,
+                                const RGWIdentityApplier &id) {
+  return out << id.to_str();
+}
```

Comment:
the to_str() interface imposes quite a bit of overhead by requiring the implementations to create a stringstream and convert it to string, just to send it back into an ostream

the MDS provides a good example of polymorphic stream output operators:
```c++
class LogEvent {
...
  virtual void print(std::ostream& out) = 0;
};
inline ostream& operator<<(ostream& out, LogEvent& le) {
  le.print(out);
  return out;
}
```
```c++
class EOpen : public LogEvent {
...
  void print(ostream& out) const {
    out << "EOpen " << metablob << ", " << inos.size() << " open files";
  }
};
```